File: backend/internal/model/lab_sheet.go
package model

import (
	"encoding/json"
	"time"
)

// LabSheetTemplate represents a reusable Univer Sheet template for a specific test item.
// Each template defines the workbook structure (rows, columns, formulas, merged cells)
// and which ranges are editable vs read-only.
type LabSheetTemplate struct {
	ID             uint            `gorm:"primarykey" json:"id"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
	Code           string          `gorm:"uniqueIndex;size:64;not null" json:"code"`
	Name           string          `gorm:"size:200;not null" json:"name"`
	TestItemID     uint            `gorm:"index" json:"test_item_id"`
	NodeCode       string          `gorm:"size:64" json:"node_code"`
	Structure      json.RawMessage `gorm:"type:jsonb;not null" json:"structure"`
	EditableRanges json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"editable_ranges"`
	ReadOnlyRanges json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"readonly_ranges"`
	Version        int             `gorm:"default:1" json:"version"`
	Status         int             `gorm:"default:1" json:"status"` // 1=active 0=disabled
}

func (LabSheetTemplate) TableName() string { return "lab_sheet_templates" }

// LabSheet represents an instance of a test sheet filled in for a specific task order.
// Created by the workflow engine when a process advances to node_data_entry,
// or manually created when a user picks a template and starts filling data.
type LabSheet struct {
	ID             uint            `gorm:"primarykey" json:"id"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
	TaskOrderID    uint            `gorm:"index" json:"task_order_id"`
	TestItemID     uint            `gorm:"index" json:"test_item_id"`
	TemplateID     *uint           `gorm:"index" json:"template_id"`
	NodeCode       string          `gorm:"size:64" json:"node_code"`
	SheetData      json.RawMessage `gorm:"type:jsonb;not null" json:"sheet_data"`
	FormulaResults json.RawMessage `gorm:"type:jsonb;default:'{}'" json:"formula_results"`
	Status         int             `gorm:"default:0;index" json:"status"` // 0=draft 1=submitted 2=reviewed 3=audited
	CreatedBy      *uint           `json:"created_by"`
	UpdatedBy      *uint           `json:"updated_by"`

	Template *LabSheetTemplate `gorm:"foreignKey:TemplateID" json:"template,omitempty"`
}

func (LabSheet) TableName() string { return "lab_sheets" }

File: frontend/src/api/business/labSheet.ts
import request from '../request'

export interface LabSheetTemplate {
  id: number
  code: string
  name: string
  test_item_id: number
  node_code: string
  structure: any
  editable_ranges: string[]
  readonly_ranges: string[]
  version: number
  status: number
}

export interface LabSheet {
  id: number
  task_order_id: number
  test_item_id: number
  template_id: number | null
  node_code: string
  sheet_data: any
  formula_results: any
  status: number
  created_by: number | null
  updated_by: number | null
  template?: LabSheetTemplate
  created_at: string
  updated_at: string
}

export function listLabSheetTemplates(params?: { name?: string; test_item_id?: string; node_code?: string; status?: string }) {
  return request.get('/lab-sheets/templates', { params })
}

export function getLabSheetTemplate(id: number) {
  return request.get(`/lab-sheets/templates/${id}`)
}

export function createLabSheetTemplate(data: Partial<LabSheetTemplate>) {
  return request.post('/lab-sheets/templates', data)
}

export function updateLabSheetTemplate(id: number, data: Partial<LabSheetTemplate>) {
  return request.put(`/lab-sheets/templates/${id}`, data)
}

export function deleteLabSheetTemplate(id: number) {
  return request.delete(`/lab-sheets/templates/${id}`)
}

export function getLabSheetTemplateByItem(testItemID: number) {
  return request.get(`/lab-sheets/templates/by-item/${testItemID}`)
}

export function listLabSheets(params?: { task_order_id?: string; test_item_id?: string; node_code?: string; status?: string }) {
  return request.get('/lab-sheets', { params })
}

export function getLabSheet(id: number) {
  return request.get(`/lab-sheets/${id}`)
}

export function createLabSheet(data: {
  task_order_id: number
  test_item_id: number
  template_id?: number | null
  node_code?: string
  sheet_data?: any
}) {
  return request.post('/lab-sheets', data)
}

export function updateLabSheet(id: number, data: {
  sheet_data?: any
  formula_results?: any
  status?: number
  node_code?: string
}) {
  return request.put(`/lab-sheets/${id}`, data)
}

export function deleteLabSheet(id: number) {
  return request.delete(`/lab-sheets/${id}`)
}